graphix/opengraph.py

```diff
@@ -0,0 +1,165 @@
+
+from __future__ import annotations
+
+import math
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+import networkx as nx
+
+from graphix.generator import generate_from_graph
+
+if TYPE_CHECKING:
+    from graphix.pattern import Pattern
+    from graphix.pauli import Plane
+
+
+@dataclass
+class Measurement:
+    """An MBQC measurement.
+
+    :param angle: the angle of the measurement. Should be between [0, 2)
+    :param plane: the measurement plane
+    """
+
+    angle: float
+    plane: Plane
+
+    def isclose(self, other: Measurement, rel_tol=1e-09, abs_tol=0.0) -> bool:
+        """Compares if two measurements have the same plane and their angles
+        are close.
+
+        Example
+        -------
+        >>> from graphix.opengraph import Measurement
+        >>> from graphix.pauli import Plane
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        True
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.YZ))
+        False
+        >>> Measurement(0.1, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        False
+        """
+        if not isinstance(other, Measurement):
+            return NotImplemented
+
+        return math.isclose(self.angle, other.angle, rel_tol=rel_tol, abs_tol=abs_tol) and self.plane == other.plane
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.opengraph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    _inside: nx.Graph
+    _inputs: list[int]
+    _outputs: list[int]
+    _meas: dict[int, Measurement]
+
+    def __eq__(self, other):
+        """Checks the two open graphs are equal
+
+        This doesn't check they are equal up to an isomorphism"""
+
+        if not nx.utils.graphs_equal(self._inside, other._inside):
+            return False
+
+        if self.inputs != other.inputs or self.outputs != other.outputs:
+            return False
+
+        if set(self.measurements.keys()) != set(other.measurements.keys()):
+            return False
+
+        return all(m.isclose(other.measurements[node]) for node, m in self.measurements.items())
+
+    def __init__(
+        self,
+        inside: nx.Graph,
+        measurements: dict[int, Measurement],
```

Comment:
Please use `Mapping` and `Sequence` if possible.

Reply:
Hey, I'm not sure I understand what you mean. Could you describe the change you'd like to see a little more?

Reply:
This is a rather general remark: if you only needs something that can be indexed by `int`, prefer `Mapping[int, Measurement]` for flexibility.
But anyway, the usage here is actually reasonable.
Sorry for confusing you :cry:

Reply:
Ah I see. I was unaware of this `collections.abc.Mappings` class. It looks useful for extensibility so I've added it in.

Reply:
5e7b2aa is not the correct way to use `Mapping`, as it will obscure the concrete type.
My implication was that, to use `Mapping` for the argument type and then cast it to `dict` inside `__init__`.
If you want to make this class itself generic, you need `TypeVar`, but it may be going too further.

Reply:
If you're interested in more details, please see https://github.com/TeamGraphix/fastflow/tree/master/python/fastflow for example.

Reply:
Oh I thought the point was to obscure over the concrete type 😅 , so that we may use any object that can be indexed with `int` and return a `Measurement`.

I see how you are using `Mapping` and converting it to a `dict` in the fastflow repo. For simplicity, I think we could stick to `dict` for now. We can always change the code to use `Mapping` in future if there is a need and it will be backwards compatible.